File: advent2024/src/main/kotlin/solutions/Day01.kt
package net.libexec.adventofcode.solutions

import net.libexec.adventofcode.Day
import net.libexec.adventofcode.util.collections.splitIntegers
import kotlin.math.abs

class Day01(loadData: Boolean = true) : Day("01", loadData) {
    override fun solveA(): String {
        val values = parseInput()
        var distance = 0

        values.forEach {
            distance += abs(it.first - it.second)
        }

        return distance.toString()
    }

    override fun solveB(): String {
        val values = parseInput()
        var similarity = 0

        values.forEach { l ->
            similarity += l.first * values.count { r -> r.second == l.first }
        }

        return similarity.toString()
    }

    private fun parseInput(): List<Pair<Int, Int>> {
        val result = ArrayList<Pair<Int, Int>>()
        val left = ArrayList<Int>()
        val right = ArrayList<Int>()

        input.forEach {
            val bits = it.splitIntegers("\\s+".toRegex())
            left.add(bits[0])
            right.add(bits[1])
        }

        val sortedLeft = left.sorted()
        val sortedRight = right.sorted()

        for (i in sortedLeft.indices) {
            result.add(Pair(sortedLeft[i], sortedRight[i]))
        }

        return result
    }
}

File: advent2024/advent2024/Day01.cs
using System.Collections.Immutable;
using System.Diagnostics.CodeAnalysis;
using Libexec.Advent;
using Libexec.Advent.Extensions;

namespace advent2024;

/// <summary>
/// 2024 day 1.
/// </summary>
/// <param name="isTest"><c>true</c> to load test data, <c>false</c> to load real data.</param>
/// <param name="fileSuffix">test file suffix.</param>
[SuppressMessage("ReSharper", "UnusedType.Global")]
public class Day01(bool isTest = false, string fileSuffix = "") : Day(1, isTest, fileSuffix)
{
    /// <inheritdoc/>
    public override object PartA()
    {
        var values = ParseInput();
        return values.Sum(value => Math.Abs(value.Item1 - value.Item2));
    }

    /// <inheritdoc/>
    public override object PartB()
    {
        var values = ParseInput();
        return values.Sum(l => l.Item1 * values.Count(r => r.Item2 == l.Item1));
    }

    private List<Tuple<int, int>> ParseInput()
    {
        var left = new List<int>();
        var right = new List<int>();

        foreach (var line in Input)
        {
            var bits = line.ParseMany<int>().ToArray();
            left.Add(bits[0]);
            right.Add(bits[1]);
        }
        
        var sortedLeft = left.ToImmutableArray().Sort();
        var sortedRight = right.ToImmutableArray().Sort();

        return sortedLeft.Select((value, index) => new Tuple<int, int>(value, sortedRight[index])).ToList();
    }
}